File: sensor-core/src/main/java/digital/lamp/Screen.java

package digital.lamp;

import android.app.KeyguardManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.IBinder;
import android.os.PowerManager;
import android.util.Log;
import digital.lamp.utils.Lamp_Sensor;

/**
 * Service that logs users' interactions with the screen
 * - on/off events
 * - locked/unlocked events
 *
 * @author denzil
 */
public class Screen extends Lamp_Sensor {

    private static String TAG = "LAMP::Screen";

    /**
     * Used to log the current screen status when starting the sensor
     */
    private static final String ACTION_LAMP_SCREEN_BOOT = "ACTION_LAMP_SCREEN_BOOT";

    /**
     * Broadcasted event: screen is on
     */
    public static final String ACTION_LAMP_SCREEN_ON = "ACTION_LAMP_SCREEN_ON";

    /**
     * Broadcasted event: screen is off
     */
    public static final String ACTION_LAMP_SCREEN_OFF = "ACTION_LAMP_SCREEN_OFF";

    /**
     * Broadcasted event: screen is locked
     */
    public static final String ACTION_LAMP_SCREEN_LOCKED = "ACTION_LAMP_SCREEN_LOCKED";

    /**
     * Broadcasted event: screen is unlocked
     */
    public static final String ACTION_LAMP_SCREEN_UNLOCKED = "ACTION_LAMP_SCREEN_UNLOCKED";

    private ScreenMonitor screenMonitor = null;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private static Screen.LAMPSensorObserver awareSensor;

    public static void setSensorObserver(Screen.LAMPSensorObserver observer) {
        awareSensor = observer;
    }

    public static Screen.LAMPSensorObserver getSensorObserver() {
        return awareSensor;
    }

    public interface LAMPSensorObserver {
        void onScreenOn();

        void onScreenOff();

        void onScreenLocked();

        void onScreenUnlocked();
    }

    @Override
    public void onCreate() {
        super.onCreate();

//        AUTHORITY = Screen_Provider.getAuthority(this);

        if (Lamp.DEBUG) Log.d(TAG, "Screen service created!");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (screenMonitor != null) unregisterReceiver(screenMonitor);

        if (Lamp.DEBUG) Log.d(TAG, "Screen service terminated...");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        super.onStartCommand(intent, flags, startId);

        if (intent != null && intent.getAction() != null) {

            PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
            KeyguardManager km = (KeyguardManager) getSystemService(KEYGUARD_SERVICE);

            if (intent.getAction().equals(ACTION_LAMP_SCREEN_ON)) {

                if (awareSensor != null) awareSensor.onScreenOn();
                Log.d(TAG, ACTION_LAMP_SCREEN_ON);
                sendBroadcast(new Intent(ACTION_LAMP_SCREEN_ON));

                if (km.isKeyguardLocked()) {
                        if (awareSensor != null) awareSensor.onScreenUnlocked();
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED);
                    sendBroadcast(new Intent(ACTION_LAMP_SCREEN_LOCKED));
                }

                return START_STICKY;
            }
            if (intent.getAction().equals(ACTION_LAMP_SCREEN_OFF)) {
                if (awareSensor != null) awareSensor.onScreenOff();

                if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_OFF);
                sendBroadcast(new Intent(ACTION_LAMP_SCREEN_OFF));

                if (km.isKeyguardLocked()) {
                    if (awareSensor != null) awareSensor.onScreenLocked();

                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED);
                    sendBroadcast(new Intent(ACTION_LAMP_SCREEN_LOCKED));
                }
                return START_STICKY;
            }

            if (intent.getAction().equals(ACTION_LAMP_SCREEN_UNLOCKED)) {
                if (!km.isKeyguardLocked()) {
                    if (awareSensor != null) awareSensor.onScreenUnlocked();

                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_UNLOCKED);
                    sendBroadcast(new Intent(ACTION_LAMP_SCREEN_UNLOCKED));
                }
                return START_STICKY;
            }
        }

        if (PERMISSIONS_OK) {
            if (screenMonitor == null) {
                screenMonitor = new ScreenMonitor();

                IntentFilter filter = new IntentFilter();
                filter.addAction(Intent.ACTION_SCREEN_ON);
                filter.addAction(Intent.ACTION_SCREEN_OFF);
                filter.addAction(Intent.ACTION_USER_PRESENT);
                filter.addAction(Screen.ACTION_LAMP_SCREEN_BOOT);
                registerReceiver(screenMonitor, filter);

                PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
                KeyguardManager km = (KeyguardManager) getSystemService(KEYGUARD_SERVICE);

                if (pm.isInteractive()) {
                    if (awareSensor != null) awareSensor.onScreenOn();

                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_ON);
                    sendBroadcast(new Intent(ACTION_LAMP_SCREEN_ON));

                    if (km.isKeyguardLocked()) {
                        if (awareSensor != null) awareSensor.onScreenUnlocked();

                        if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED);
                        sendBroadcast(new Intent(ACTION_LAMP_SCREEN_LOCKED));
                    } else {
                        if (awareSensor != null) awareSensor.onScreenUnlocked();

                        if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_UNLOCKED);
                        sendBroadcast(new Intent(ACTION_LAMP_SCREEN_UNLOCKED));
                    }
                } else {
                    if (awareSensor != null) awareSensor.onScreenOff();

                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_OFF);
                    sendBroadcast(new Intent(ACTION_LAMP_SCREEN_OFF));

                    if (km.isKeyguardLocked()) {
                            if (awareSensor != null) awareSensor.onScreenLocked();

                        if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED);
                        sendBroadcast(new Intent(ACTION_LAMP_SCREEN_LOCKED));
                    }
                }
            }

            if (Lamp.DEBUG) Log.d(TAG, "Screen service active...");
        }
        return START_STICKY;
    }

    private class ScreenMonitor extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
            KeyguardManager km = (KeyguardManager) getSystemService(KEYGUARD_SERVICE);

            if (intent.getAction().equals(Intent.ACTION_SCREEN_ON)) {
                startService(new Intent(context, Screen.class).setAction(Screen.ACTION_LAMP_SCREEN_ON));
            }
            if (intent.getAction().equals(Intent.ACTION_SCREEN_OFF) || !pm.isInteractive()) {
                startService(new Intent(context, Screen.class).setAction(Screen.ACTION_LAMP_SCREEN_OFF));
            }
            if (intent.getAction().equals(Intent.ACTION_USER_PRESENT) && !km.isKeyguardLocked()) {
                startService(new Intent(context, Screen.class).setAction(Screen.ACTION_LAMP_SCREEN_UNLOCKED));
            }
        }
    }

}

File: sensor-core/src/main/java/digital/lamp/Screen.kt
package digital.lamp

import digital.lamp.Screen.ScreenMonitor
import android.content.Intent
import android.os.IBinder
import android.os.PowerManager
import android.app.KeyguardManager
import android.app.Service
import android.content.IntentFilter
import android.content.BroadcastReceiver
import android.content.Context
import android.util.Log

/**
 * Service that logs users' interactions with the screen
 * - on/off events
 * - locked/unlocked events
 *
 * @author denzil
 */
class Screen : Service() {
    private var screenMonitor: ScreenMonitor? = null
    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    interface LAMPSensorObserver {
        fun onScreenOn()
        fun onScreenOff()
        fun onScreenLocked()
        fun onScreenUnlocked()
    }

    override fun onCreate() {
        super.onCreate()
        if (Lamp.DEBUG) Log.d(TAG, "Screen service created!")
    }

    override fun onDestroy() {
        super.onDestroy()
        if (screenMonitor != null) unregisterReceiver(screenMonitor)
        if (Lamp.DEBUG) Log.d(TAG, "Screen service terminated...")
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        if (intent.action != null) {
            val km = getSystemService(KEYGUARD_SERVICE) as KeyguardManager
            if (intent.action == ACTION_LAMP_SCREEN_ON) {
                if (sensorObserver != null) sensorObserver!!.onScreenOn()
                Log.d(TAG, ACTION_LAMP_SCREEN_ON)
                sendBroadcast(Intent(ACTION_LAMP_SCREEN_ON))
                if (km.isKeyguardLocked) {
                    if (sensorObserver != null) sensorObserver!!.onScreenUnlocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_LOCKED))
                }
                return START_STICKY
            }
            if (intent.action == ACTION_LAMP_SCREEN_OFF) {
                if (sensorObserver != null) sensorObserver!!.onScreenOff()
                if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_OFF)
                sendBroadcast(Intent(ACTION_LAMP_SCREEN_OFF))
                if (km.isKeyguardLocked) {
                    if (sensorObserver != null) sensorObserver!!.onScreenLocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_LOCKED))
                }
                return START_STICKY
            }
            if (intent.action == ACTION_LAMP_SCREEN_UNLOCKED) {
                if (!km.isKeyguardLocked) {
                    if (sensorObserver != null) sensorObserver!!.onScreenUnlocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_UNLOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_UNLOCKED))
                }
                return START_STICKY
            }
        }
        if (screenMonitor == null) {
            screenMonitor = ScreenMonitor()
            val filter = IntentFilter()
            filter.addAction(Intent.ACTION_SCREEN_ON)
            filter.addAction(Intent.ACTION_SCREEN_OFF)
            filter.addAction(Intent.ACTION_USER_PRESENT)
            filter.addAction(ACTION_LAMP_SCREEN_BOOT)
            registerReceiver(screenMonitor, filter)
            val pm = getSystemService(POWER_SERVICE) as PowerManager
            val km = getSystemService(KEYGUARD_SERVICE) as KeyguardManager
            if (pm.isInteractive) {
                if (sensorObserver != null) sensorObserver!!.onScreenOn()
                if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_ON)
                sendBroadcast(Intent(ACTION_LAMP_SCREEN_ON))
                if (km.isKeyguardLocked) {
                    if (sensorObserver != null) sensorObserver!!.onScreenUnlocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_LOCKED))
                } else {
                    if (sensorObserver != null) sensorObserver!!.onScreenUnlocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_UNLOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_UNLOCKED))
                }
            } else {
                if (sensorObserver != null) sensorObserver!!.onScreenOff()
                if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_OFF)
                sendBroadcast(Intent(ACTION_LAMP_SCREEN_OFF))
                if (km.isKeyguardLocked) {
                    if (sensorObserver != null) sensorObserver!!.onScreenLocked()
                    if (Lamp.DEBUG) Log.d(TAG, ACTION_LAMP_SCREEN_LOCKED)
                    sendBroadcast(Intent(ACTION_LAMP_SCREEN_LOCKED))
                }
            }
        }
        if (Lamp.DEBUG) Log.d(TAG, "Screen service active...")
        return START_STICKY
    }

    private inner class ScreenMonitor : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val pm = getSystemService(POWER_SERVICE) as PowerManager
            val km = getSystemService(KEYGUARD_SERVICE) as KeyguardManager
            if (intent.action == Intent.ACTION_SCREEN_ON) {
                startService(Intent(context, Screen::class.java).setAction(ACTION_LAMP_SCREEN_ON))
            }
            if (intent.action == Intent.ACTION_SCREEN_OFF || !pm.isInteractive) {
                startService(Intent(context, Screen::class.java).setAction(ACTION_LAMP_SCREEN_OFF))
            }
            if (intent.action == Intent.ACTION_USER_PRESENT && !km.isKeyguardLocked) {
                startService(Intent(context, Screen::class.java).setAction(ACTION_LAMP_SCREEN_UNLOCKED))
            }
        }
    }

    companion object {
        private const val TAG = "LAMP::Screen"

        /**
         * Used to log the current screen status when starting the sensor
         */
        private const val ACTION_LAMP_SCREEN_BOOT = "ACTION_LAMP_SCREEN_BOOT"

        /**
         * Broadcasted event: screen is on
         */
        const val ACTION_LAMP_SCREEN_ON = "ACTION_LAMP_SCREEN_ON"

        /**
         * Broadcasted event: screen is off
         */
        const val ACTION_LAMP_SCREEN_OFF = "ACTION_LAMP_SCREEN_OFF"

        /**
         * Broadcasted event: screen is locked
         */
        const val ACTION_LAMP_SCREEN_LOCKED = "ACTION_LAMP_SCREEN_LOCKED"

        /**
         * Broadcasted event: screen is unlocked
         */
        const val ACTION_LAMP_SCREEN_UNLOCKED = "ACTION_LAMP_SCREEN_UNLOCKED"
        var sensorObserver: LAMPSensorObserver? = null
    }
}